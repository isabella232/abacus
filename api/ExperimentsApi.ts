import _ from 'lodash'
import * as yup from 'yup'

import {
  ExperimentBare,
  experimentBareSchema,
  ExperimentFull,
  ExperimentFullNew,
  experimentFullNewOutboundSchema,
  experimentFullNewSchema,
  experimentFullSchema,
  MetricAssignmentNew,
  metricAssignmentNewOutboundSchema,
  metricAssignmentNewSchema,
  Status,
  yupPick,
} from '@/lib/schemas'

import { fetchApi } from './utils'

/**
 * Attempts to create a new experiment.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newExperiment: ExperimentFullNew) {
  const validatedNewExperiment = await experimentFullNewSchema.validate(newExperiment, { abortEarly: false })
  const outboundNewExperiment = experimentFullNewOutboundSchema.cast(validatedNewExperiment)
  const returnedExperiment = await fetchApi('POST', '/experiments', outboundNewExperiment)
  return await experimentFullSchema.validate(returnedExperiment)
}

/**
 * Attempts to PUT an experiment, overwriting the existing experiment.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(experimentId: number, experiment: ExperimentFullNew) {
  const validatedExperiment = await experimentFullNewSchema.validate(experiment, { abortEarly: false })
  const outboundExperiment = experimentFullNewOutboundSchema.cast(validatedExperiment)
  const returnedExperiment = await fetchApi('PUT', `/experiments/${experimentId}`, outboundExperiment)
  return await experimentFullSchema.validate(returnedExperiment)
}

/**
 * Attempts to patch an experiment.
 *
 * Doesn't work with nested fields.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function patch(experimentId: number, experimentPatch: Partial<ExperimentFull>) {
  // We dynamically construct a schema for validation, but we do this simply and shallowly
  const dynamicValidationSchema = yupPick(experimentFullSchema, Object.keys(experimentPatch))
  const validatedExperimentPatch = await dynamicValidationSchema.validate(experimentPatch, { abortEarly: false })

  // Similarly we dynamically construct a schema for outbound casting
  const dynamicOutboundCastSchema = yupPick(experimentFullSchema, Object.keys(experimentPatch)).snakeCase()
  const outboundExperimentPatch = dynamicOutboundCastSchema.cast(validatedExperimentPatch)

  const returnedExperiment = await fetchApi('PATCH', `/experiments/${experimentId}`, outboundExperimentPatch)
  return await experimentFullSchema.validate(returnedExperiment)
}

async function changeStatus(experimentId: number, status: Status) {
  await fetchApi('PUT', `/experiments/${experimentId}/status`, { status })
}

/**
 * Attempts to assign a metric to an experiment.
 *
 * Metric Assignment is a bit funky right now, so we do it here.
 * In particular it require passing in all existing metricAssignments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @param experiment A full experiment object, needed to fill in the existing metricAssignments, will be removed in the future.
 * @param metricAssignment The new metricAssignment.
 */
async function assignMetric(experiment: ExperimentFull, metricAssignment: MetricAssignmentNew) {
  const validatedMetricAssignment = await metricAssignmentNewSchema.validate(metricAssignment)
  const outboundMetricAssignment = metricAssignmentNewOutboundSchema.cast(validatedMetricAssignment)

  // The backend replaces all the metricAssignments, so we format it as metricAssignmentNewOutbound.
  const outboundExistingMetricAssignments = yup
    .array(metricAssignmentNewOutboundSchema)
    .cast(experiment.metricAssignments.map((metricAssignment) => _.omit(metricAssignment, 'metricAssignmentId')))
  // istanbul ignore next; Shouldn't occur
  if (!outboundExistingMetricAssignments) {
    throw new Error('Something went wrong while transforming existing metricAssignments to outbound form.')
  }

  const metricAssignments = [...outboundExistingMetricAssignments, outboundMetricAssignment]

  const returnedExperiment = await fetchApi('PATCH', `/experiments/${experiment.experimentId}`, {
    metric_assignments: metricAssignments,
  })
  return await experimentFullSchema.validate(returnedExperiment)
}

/**
 * Finds all the available experiments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<ExperimentBare[]> {
  const { experiments } = await fetchApi('GET', '/experiments')
  return await yup.array(experimentBareSchema).defined().validate(experiments, { abortEarly: false })
}

/**
 * Fetches the experiment with full details.
 *
 * @param id - The ID of the experiment to fetch.
 */
async function findById(id: number): Promise<ExperimentFull> {
  const experiment = await fetchApi('GET', `/experiments/${id}`)
  return await experimentFullSchema.validate(experiment, { abortEarly: false })
}

const ExperimentsApi = {
  create,
  put,
  patch,
  assignMetric,
  changeStatus,
  findAll,
  findById,
}

export default ExperimentsApi
