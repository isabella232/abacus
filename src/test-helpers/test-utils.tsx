/* eslint-disable @typescript-eslint/no-unsafe-return,@typescript-eslint/explicit-module-boundary-types */
import { act, fireEvent, Queries, render as actualRender, RenderOptions, screen } from '@testing-library/react'
import { StaticRouter } from 'react-router'
import mediaQuery from 'css-mediaquery'
import { Formik, FormikValues } from 'formik'
import React from 'react'
import ThemeProvider from 'src/styles/ThemeProvider'
import { ValidationError } from 'yup'

/**
 * A wrapped unit-test react-renderer, useful for adding React Contexts globally.
 */
export const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender(
    (
      <StaticRouter>
        <ThemeProvider>{ui}</ThemeProvider>
      </StaticRouter>
    ) as React.ReactElement,
    options,
  ) as ReturnType<typeof actualRender>

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
export function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}

/**
 * Mock Formik for rendering Formik components when you don't care about the formik connection.
 */
export const MockFormik = ({
  children,
  initialValues = {},
}: {
  children: React.ReactNode
  initialValues?: FormikValues
}) => {
  return (
    <Formik initialValues={initialValues} onSubmit={() => undefined}>
      {children}
    </Formik>
  )
}

/**
 * Validation Error Displayer
 */
export async function validationErrorDisplayer<T>(promise: Promise<T>): Promise<T> {
  try {
    return await promise
  } catch (err) {
    if (err instanceof ValidationError) {
      expect(err.errors).toEqual([])
    }
    throw err
  }
}

/**
 * Change the value in a form field.
 */
export async function changeFieldByRole(role: string, name: RegExp, value: string) {
  const field = screen.getByRole(role, { name: name })
  // eslint-disable-next-line @typescript-eslint/require-await
  await act(async () => {
    fireEvent.change(field, { target: { value: value } })
  })
}
