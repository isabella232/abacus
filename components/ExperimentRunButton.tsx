import { Button, Dialog, DialogActions, DialogContent, Tooltip, Typography } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { useSnackbar } from 'notistack'
import React, { useState } from 'react'

import ExperimentsApi from '@/api/ExperimentsApi'
import { ExperimentFull, Status } from '@/lib/schemas'

import LoadingButtonContainer from './LoadingButtonContainer'

const useStyles = makeStyles((_theme: Theme) =>
  createStyles({
    root: {},
  }),
)

const ExperimentRunButton = ({
  experiment,
  experimentReloadRef,
}: {
  experiment: ExperimentFull | null
  experimentReloadRef: React.MutableRefObject<() => void>
}) => {
  const _classes = useStyles()

  const { enqueueSnackbar } = useSnackbar()

  const canRunExperiment = experiment && experiment.status === Status.Staging
  const [isAskingToConfirmRunExperiment, setIsAskingToConfirmRunExperiment] = useState<boolean>(false)
  const onAskToConfirmRunExperiment = () => setIsAskingToConfirmRunExperiment(true)
  const onCancelRunExperiment = () => setIsAskingToConfirmRunExperiment(false)
  const [isSubmittingRunExperiment, setIsSubmittingRunExperiment] = useState<boolean>(false)
  const onConfirmRunExperiment = async () => {
    try {
      // istanbul ignore next; Shouldn't occur
      if (!experiment) {
        throw Error('Missing experiment, this should not happen')
      }

      setIsSubmittingRunExperiment(true)
      await ExperimentsApi.changeStatus(experiment.experimentId, Status.Running)
      enqueueSnackbar('Experiment Running!', { variant: 'success' })
      experimentReloadRef.current()
      setIsAskingToConfirmRunExperiment(false)
    } catch (e) /* istanbul ignore next; Shouldn't occur */ {
      console.log(e)
      enqueueSnackbar('Oops! Something went wrong while trying to run your experiment.', { variant: 'error' })
    } finally {
      setIsSubmittingRunExperiment(false)
    }
  }

  return (
    <>
      <Tooltip title={canRunExperiment ? '' : `This experiment is ${experiment?.status}.`}>
        <span>
          <Button
            variant='outlined'
            color='secondary'
            disabled={!canRunExperiment}
            onClick={onAskToConfirmRunExperiment}
          >
            Run
          </Button>
        </span>
      </Tooltip>
      <Dialog open={isAskingToConfirmRunExperiment} aria-labelledby='confirm-run-experiment-dialog-title'>
        <DialogContent>
          <Typography variant='body1'>Are you sure you want to run this experiment?</Typography>
        </DialogContent>
        <DialogActions>
          <Button onClick={onCancelRunExperiment}>Cancel</Button>
          <LoadingButtonContainer isLoading={isSubmittingRunExperiment}>
            <Button
              variant='contained'
              color='primary'
              disabled={isSubmittingRunExperiment}
              onClick={onConfirmRunExperiment}
            >
              Run
            </Button>
          </LoadingButtonContainer>
        </DialogActions>
      </Dialog>
    </>
  )
}

export default ExperimentRunButton
