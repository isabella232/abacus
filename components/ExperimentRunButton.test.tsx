/* eslint-disable @typescript-eslint/require-await */
import { fireEvent, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import MockDate from 'mockdate'
import * as notistack from 'notistack'
import React from 'react'

import ExperimentsApi from '@/api/ExperimentsApi'
import Fixtures from '@/test-helpers/fixtures'
import { render } from '@/test-helpers/test-utils'

import ExperimentRunButton from './ExperimentRunButton'

MockDate.set('2020-07-21')

jest.mock('@/api/ExperimentsApi')
const mockedExperimentsApi = ExperimentsApi as jest.Mocked<typeof ExperimentsApi>

jest.mock('notistack')
const mockedNotistack = notistack as jest.Mocked<typeof notistack>
mockedNotistack.useSnackbar.mockImplementation(() => ({
  enqueueSnackbar: jest.fn(),
  closeSnackbar: jest.fn(),
}))

test('renders as expected', () => {
  const experimentReloadRef: React.MutableRefObject<() => void> = { current: jest.fn() }
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentRunButton {...{ experiment, experimentReloadRef }} />)

  expect(container).toMatchSnapshot()
})

test('runs an experiment', async () => {
  const experimentReloadRef: React.MutableRefObject<() => void> = { current: jest.fn() }
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentRunButton {...{ experiment, experimentReloadRef }} />)

  mockedExperimentsApi.changeStatus.mockReset()
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  mockedExperimentsApi.changeStatus.mockImplementationOnce(async () => experiment)

  const firstRunButton = screen.getByRole('button', { name: /Run/ })

  // First Opening - We cancel
  fireEvent.click(firstRunButton)

  await waitFor(() => screen.getByRole('button', { name: /Cancel/ }))

  expect(container).toMatchSnapshot()

  const cancelButton = screen.getByRole('button', { name: /Cancel/ })
  fireEvent.click(cancelButton)
  await waitForElementToBeRemoved(cancelButton)

  expect(mockedExperimentsApi.changeStatus).toHaveBeenCalledTimes(0)
  expect(experimentReloadRef.current).toHaveBeenCalledTimes(0)

  // Second Opening - We run
  fireEvent.click(firstRunButton)

  await waitFor(() => screen.getByRole('button', { name: /Cancel/ }))
  const cancelButton2nd = screen.getByRole('button', { name: /Cancel/ })

  const allRunButtons = screen.getAllByRole('button', { name: /Run/ })
  allRunButtons.forEach((button) => fireEvent.click(button))

  await waitForElementToBeRemoved(cancelButton2nd)

  expect(mockedExperimentsApi.changeStatus).toHaveBeenCalledTimes(1)
  expect(experimentReloadRef.current).toHaveBeenCalledTimes(1)
  expect(mockedExperimentsApi.changeStatus).toMatchInlineSnapshot(`
    [MockFunction] {
      "calls": Array [
        Array [
          1,
          "running",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  `)
})
