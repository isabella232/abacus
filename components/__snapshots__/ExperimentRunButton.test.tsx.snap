// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <span
    class=""
    title=""
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Run
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </span>
</div>
`;

exports[`runs an experiment 1`] = `
<div
  aria-hidden="true"
>
  <span
    class=""
    title=""
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Run
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </span>
</div>
`;
